espnet2/enh/separator/ineube_separator.py

```diff
@@ -0,0 +1,147 @@
+
+        if window == "hanning":
+            window = torch.hann_window(n_fft)
+        elif window == "hamming":
+            window = torch.hamming_window(n_fft)
+        elif window is None:
+            pass
+        else:
+            raise NotImplementedError
+        fft_c_channels = n_fft // 2 + 1
+        self.enc, self.dec = make_enc_dec(
+            "torch_stft", n_fft, n_fft, stride, window=window
+        )
+        self.dnn1 = TCNDenseUNet(
+            n_spk,
+            fft_c_channels,
+            mic_channels,
+            hid_chans,
+            hid_chans_dense,
+            ksz_dense,
+            ksz_tcn,
+            tcn_repeats,
+            tcn_blocks,
+            tcn_channels,
+            activation=activation,
+        )
+
+        self.dnn2 = TCNDenseUNet(
+            n_spk,
+            fft_c_channels,
+            mic_channels + 2,
+            hid_chans,
+            hid_chans_dense,
+            ksz_dense,
+            ksz_tcn,
+            tcn_repeats,
+            tcn_blocks,
+            tcn_channels,
+            activation=activation,
+        )
+
+    def unfold(self, tf_rep, chunk_size):
+        bsz, freq, _ = tf_rep.shape
+
+        est_unfolded = torch.nn.functional.unfold(
+            torch.nn.functional.pad(
+                tf_rep, (chunk_size, chunk_size), mode="constant"
+            ).unsqueeze(-1),
+            kernel_size=(2 * chunk_size + 1, 1),
+            padding=(0, 0),
+            stride=(1, 1),
+        )
+        n_chunks = est_unfolded.shape[-1]
+        est_unfolded = est_unfolded.reshape(bsz, freq, 2 * chunk_size + 1, n_chunks)
+        est_unfolded = est_unfolded.transpose(1, 2)
+        return est_unfolded
+
+    def mfmcwf(self, mixture, estimate, n_chunks):
+        bsz, mics, _, frames = mixture.shape
+        mix_unfolded = self.unfold(
+            mixture.reshape(bsz * mics, -1, frames), n_chunks
+        ).reshape(bsz, mics * (2 * n_chunks + 1), -1, frames)
+
+        mix_unfolded = af_transforms.to_torch_complex(mix_unfolded.double())
+        estimate1 = af_transforms.to_torch_complex(estimate.double())
+        zeta = torch.einsum("bmft, bft->bmf", mix_unfolded, estimate1.conj())
+        scm_mix = torch.einsum("bmft, bnft->bmnf", mix_unfolded, mix_unfolded.conj())
+        inv_scm_mix = torch.inverse(
+            condition_scm(scm_mix.permute(0, 3, 1, 2), 1e-4)
+        ).permute(0, 2, 3, 1)
+        bf_vector = torch.einsum("bmnf, bnf->bmf", inv_scm_mix, zeta)
+
+        beamformed = torch.einsum("...mf,...mft->...ft", bf_vector.conj(), mix_unfolded)
+        beamformed = af_transforms.from_torch_complex(beamformed).float()
+
+        return beamformed
+
+    def forward(
+        self,
+        input: Union[torch.Tensor, ComplexTensor],
+        ilens: torch.Tensor,
+        output_from="dnn1",
+        n_chunks=3,
+        freeze_dnn1=False,
+    ) -> Tuple[List[Union[torch.Tensor, ComplexTensor]], torch.Tensor, OrderedDict]:
```

Comment:
On second thought, I think it is better to move the arguments `output_from="dnn1"`, `n_chunks=3`, and `freeze_dnn1=False` to `__init__`, because it is messy to add model-specific argument in the current common [espnet_model](https://github.com/espnet/espnet/blob/master/espnet2/enh/espnet_model.py#L140).

In my recent trial, I find it convenient enough to move these arguments to model attributes, so that
* training of dnn1 or dnn2 can be configured directly in a config file
* it is more clear to understand the model behavior through the model config

---

What I did is as follows:
1. remove arguments `output_from`, `n_chunks`, and `freeze_dnn1`.
2. add attributes `train_dnn1`,  `train_dnn2`, and `n_chunks` in `__init__`.
3. add some code in `__init__` to freeze parameter accordingly:
   > ```python
   > if not self.train_dnn1:
   >    # freeze dnn1
   > elif not self.train_dnn1:
   >    # freeze dnn2
   > ```
4. modify the logic in `forward` to:
   > ```python
   > if self.train_dnn2:
   >    # same as output_from=="dnn2"
   > elif self.train_dnn1:
   >    # same as output_form=="dnn1"
   > else:
   >    # same as output_form=="beam"
   > ```